### ISSUE TITLE
a rule to find redundant slice variable declarations

### ISSUE THREAD

**Issue by alexandear** [mgechev/revive#1704]
**Is your feature request related to a problem? Please describe.**

After refactoring, sometimes there are `nil` slices declarations `var x []T` that are passed to some calls `(x...)`. These slices have no effect and can be removed.

**Describe the solution you'd like**

A new rule that flags:

- Local declarations like `var x []T` with no initializer.
- Where every usage of x is only as a variadic expansion in calls `(x...)`.
- These are reported as no-op declarations because the zero value is nil, and expanding a ` slice passes zero arguments.

**Describe alternatives you've considered**

N/A

**Additional context**

Inspired by https://github.com/kubernetes-sigs/kueue/pull/10324

<details><summary>Copilot writes an analyzer for this</summary>
<p>

```go
package main

import (
	"go/ast"
	"go/token"
	"go/types"

	"golang.org/x/tools/go/analysis"
	"golang.org/x/tools/go/analysis/singlechecker"
)

var analyzer = &analysis.Analyzer{
	Name: "noopvariadicslice",
	Doc:  "reports local slice var declarations that only expand nil variadic arguments",
	Run:  run,
}

func main() {
	singlechecker.Main(analyzer)
}

type candidate struct {
	ident *ast.Ident
	obj   *types.Var
}

func run(pass *analysis.Pass) (any, error) {
	parents := buildParentMap(pass.Files)
	candidates := findCandidates(pass)

	for _, c := range candidates {
		uses := 0
		onlyVariadicExpansionUses := true

		for ident, obj := range pass.TypesInfo.Uses {
			if obj != c.obj {
				continue
			}
			uses++
			if !isVariadicExpansionUse(ident, parents) {
				onlyVariadicExpansionUses = false
				break
			}
		}

		if uses == 0 || !onlyVariadicExpansionUses {
			continue
		}

		pass.Reportf(
			c.ident.Pos(),
			"slice variable %q is nil by default and only used as variadic expansion; declaration has no effect",
			c.ident.Name,
		)
	}

	return nil, nil
}

func findCandidates(pass *analysis.Pass) []candidate {
	candidates := make([]candidate, 0)

	for _, file := range pass.Files {
		ast.Inspect(file, func(n ast.Node) bool {
			genDecl, ok := n.(*ast.GenDecl)
			if !ok || genDecl.Tok != token.VAR {
				return true
			}

			for _, spec := range genDecl.Specs {
				valueSpec, ok := spec.(*ast.ValueSpec)
				if !ok || len(valueSpec.Values) != 0 || !isSliceType(valueSpec.Type) {
					continue
				}

				for _, name := range valueSpec.Names {
					if name.Name == "_" {
						continue
					}

					obj, ok := pass.TypesInfo.Defs[name].(*types.Var)
					if !ok || obj == nil {
						continue
					}

					if obj.Parent() == pass.Pkg.Scope() {
						continue
					}

					candidates = append(candidates, candidate{ident: name, obj: obj})
				}
			}

			return true
		})
	}

	return candidates
}

func isSliceType(expr ast.Expr) bool {
	arrayType, ok := expr.(*ast.ArrayType)
	if !ok {
		return false
	}

	return arrayType.Len == nil
}

func isVariadicExpansionUse(ident *ast.Ident, parents map[ast.Node]ast.Node) bool {
	parent, ok := parents[ident]
	if !ok {
		return false
	}

	callExpr, ok := parent.(*ast.CallExpr)
	if !ok || !callExpr.Ellipsis.IsValid() || len(callExpr.Args) == 0 {
		return false
	}

	lastArg := unparen(callExpr.Args[len(callExpr.Args)-1])

	return lastArg == ident
}

func unparen(expr ast.Expr) ast.Expr {
	for {
		parenExpr, ok := expr.(*ast.ParenExpr)
		if !ok {
			return expr
		}
		expr = parenExpr.X
	}
}

func buildParentMap(files []*ast.File) map[ast.Node]ast.Node {
	parents := make(map[ast.Node]ast.Node)

	for _, file := range files {
		stack := make([]ast.Node, 0)
		ast.Inspect(file, func(n ast.Node) bool {
			if n == nil {
				stack = stack[:len(stack)-1]
				return false
			}

			if len(stack) > 0 {
				parents[n] = stack[len(stack)-1]
			}

			stack = append(stack, n)
			return true
		})
	}

	return parents
}
```

</p>
</details>

**Comment by ccoVeille** [comment]
Isn't something unparam linter reports ?
